File: src/tree/TreeSerialized.java
package tree;

import java.util.*;
import java.util.stream.Collectors;

// https://leetcode-cn.com/problems/serialize-and-deserialize-binary-tree/
public class TreeSerialized {
    static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            val = x;
        }
    }

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        return serializeByBFS(root);
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        return deserializeByBFS(data);
    }

    public static String serializeByBFS(TreeNode root) {
        StringBuilder sbl = new StringBuilder();
        if (root == null) {
            sbl.append("None,");
        } else {
            Queue<TreeNode> queue = new ArrayDeque<>();
            queue.offer(root);
            while (!queue.isEmpty()) {
                TreeNode node = queue.poll();
                if (node.left != null) {
                    queue.offer(node.left);
                    sbl.append(String.valueOf(node.val).concat(","));
                } else {
                    sbl.append("None,");
                }
                if (node.right != null) {
                    queue.offer(node.right);
                    sbl.append(String.valueOf(node.val).concat(","));
                } else {
                    sbl.append("None,");
                }
            }
        }
        return sbl.toString();
    }

    public static TreeNode deserializeByBFS(String data) {
        if ("None,".equals(data)) return null;

        Queue<String> strQueue = Arrays.stream(data.split(","))
                .collect(Collectors.toCollection(ArrayDeque::new));
        Queue<TreeNode> nodeQueue = new ArrayDeque<>();
        TreeNode root = genNode(strQueue.poll());
        nodeQueue.offer(root);
        while (!nodeQueue.isEmpty()) {
            TreeNode node = nodeQueue.poll();
            node.left = genNode(strQueue.poll());
            node.right = genNode(strQueue.poll());
            if (node.left != null) nodeQueue.offer(node.left);
            if (node.right != null) nodeQueue.offer(node.right);
        }
        return root;
    }

    public static TreeNode genNode(String str) {
        return "None".equals(str) ? null : new TreeNode(Integer.parseInt(str));
    }

    public static String serializeByDFS(TreeNode root) {
        StringBuilder sbl = new StringBuilder();
        presNode(root, sbl);
        return sbl.toString();
    }

    public static void presNode(TreeNode head, StringBuilder sbl) {
        if (head == null) {
            sbl.append("None,");
        } else {
            sbl.append(String.valueOf(head.val).concat(","));
            presNode(head.left, sbl);
            presNode(head.right, sbl);
        }
    }

    public static TreeNode deserializeByDFS(String data) {
        if ("None,".equals(data)) return null;

        ArrayDeque<String> charsDeque = Arrays.stream(data.split(","))
                .collect(Collectors.toCollection(ArrayDeque::new));
        return dePresNode(charsDeque);
    }

    public static TreeNode dePresNode(Queue<String> deque) {
        String value = deque.poll();
        if ("None".equals(value)) return null;
        TreeNode node = new TreeNode(Integer.valueOf(value));
        node.left = dePresNode(deque);
        node.right = dePresNode(deque);
        return node;
    }
}

File: src/tree/tree_serialized.py
# Definition for a binary tree node.
from collections import deque


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Codec:
    def serialize(self, root):
        """Encodes a tree to a single string.

        :type root: TreeNode
        :rtype: str
        """
        return self.level_serialize(root)

    def deserialize(self, data):
        """Decodes your encoded data to tree.

        :type data: str
        :rtype: TreeNode
        """
        return self.level_deserialize(data)

    def pre_serialize(self, root):
        """按照二叉树先序, 序列化"""
        ans = []
        self.pres(root, ans)
        return ",".join(ans)

    def pres(self, node, ans):
        if not node:
            # `.` 表示空节点
            ans.append(".")
        else:
            ans.append(str(node.val))
            self.pres(node.left, ans)
            self.pres(node.right, ans)

    def pre_deserialize(self, data):
        """根据字符串, 生成二叉树"""
        if not data:
            return None
        return self.pred(list(reversed(data.split(","))))

    def pred(self, data_list):
        """根据data字符串生成二叉树"""
        val = data_list.pop()
        if not val:
            return None
        head = TreeNode(int(val))
        head.left = self.pred(data_list)
        head.right = self.pred(data_list)
        return head

    def level_serialize(self, root):
        """按层序列化二叉树"""
        ans = []
        if root is None:
            ans.append(".")
        else:
            q = deque()
            q.append(root)
            ans.append(str(root.val))
            while q:
                node = q.popleft()
                if node.left is not None:
                    ans.append(str(node.left.val))
                    q.append(node.left)
                else:
                    ans.append(".")
                if node.right is not None:
                    ans.append(str(node.right.val))
                    q.append(node.right)
                else:
                    ans.append(".")
        return ",".join(ans)

    def level_deserialize(self, data):
        """根据字符串(按层遍历)反序列化二叉树"""

        def gen_node(s):
            return TreeNode(int(s)) if s != "." else None

        data_list = deque(data.split(","))
        if not data_list:
            return None
        head = gen_node(data_list.popleft())
        # bfs queue
        q = deque()
        if head is not None:
            q.append(head)
        while q:
            node = q.popleft()
            node.left = gen_node(data_list.popleft())
            node.right = gen_node(data_list.popleft())
            if node.left is not None:
                q.append(node.left)
            if node.right is not None:
                q.append(node.right)
        return head
